torch/utils/data/decorator.py

```diff
@@ -94,3 +102,34 @@ def deterministic_wrapper_fn(self, *args, **kwargs) -> IterDataPipe:
                             "for this DataPipe if that is acceptable for your application"
                             .format(self.cls.__name__))  # type: ignore
         return self.cls(*args, **kwargs)  # type: ignore
+
+
+######################################################
+# typing
+######################################################
+# Construct-time checking
+# Validate each DataPipe with hint as a subtype of the hint.
+def construct_time_validation(f):
+    if f.__name__ not in ('__init__', '__new__'):
+        raise TypeError("Can not decorate function {} with 'construct_time_validation'"
+                        .format(f.__name__))
+    signature = inspect.signature(f)
+    hints = get_type_hints(f)
+
+    @wraps(f)
+    def wrapper(*args, **kwargs):
+        bound = signature.bind(*args, **kwargs)
+        for argument_name, value in bound.arguments.items():
+            if argument_name in hints and isinstance(hints[argument_name], _DataPipeMeta):
+                hint = hints[argument_name]
+                if not isinstance(value, IterDataPipe):
```

Comment:
super nit: Need to add MapDataPipe support eventually.

Reply:
I will add MapDataPipe support in the short future, as I need to move metaclass to `Dataset` to support typing system for MapDataPipe. And, it will take longer time to resolve all the internal breaks.